src/savers/tvg/tvgTvgSaver.cpp

```diff
@@ -0,0 +1,210 @@
+    if (!buffer) {
+        reserved = 0;
+        return false;
+    }
+    bufferPosition = buffer;
+
+    return saveHeader();
+}
+
+
+bool TvgSaver::write()
+{
+    if (!buffer) return false;
+
+    TaskScheduler::request(this);
+
+    return true;
+}
+
+
+void TvgSaver::run(unsigned tid)
+{
+    root->serialize();  //MGS temp
+
+    ofstream outFile;
+    outFile.open(filePath, ios::out | ios::trunc | ios::binary);
+    if (!outFile.is_open()) return;
+    outFile.write(buffer,size);
+    outFile.close();
+
+    if (buffer) free(buffer);
+    buffer = nullptr;
+    bufferPosition = nullptr;
+    size = 0;
+    reserved = 0;
+};
+
+
+bool TvgSaver::close()
+{
+    this->done();
+
+    if (buffer) free(buffer);
+    buffer = nullptr;
+    bufferPosition = nullptr;
+    size = 0;
+    reserved = 0;
+
+    return true;
+}
+
+
+void TvgSaver::saveMemberIndicator(IndicatorType ind)
+{
+    if (size + sizeof(IndicatorType) > reserved) resizeBuffer(size + sizeof(IndicatorType));
+
+    memcpy(bufferPosition, &ind, sizeof(IndicatorType));
+    bufferPosition += sizeof(IndicatorType);
+    size += sizeof(IndicatorType);
+}
+
+
+void TvgSaver::saveMemberDataSize(ByteCounter byteCnt)
+{
+    if (size + sizeof(ByteCounter) > reserved) resizeBuffer(size + sizeof(ByteCounter));
+
+    memcpy(bufferPosition, &byteCnt, sizeof(ByteCounter));
+    bufferPosition += sizeof(ByteCounter);
+    size += sizeof(ByteCounter);
+}
+
+
+void TvgSaver::saveMemberDataSizeAt(ByteCounter byteCnt)
+{
+    memcpy(bufferPosition - byteCnt - sizeof(ByteCounter), &byteCnt, sizeof(ByteCounter));
+}
+
+
+void TvgSaver::skipMemberDataSize()
+{
+    if (size + sizeof(ByteCounter) > reserved) resizeBuffer(size + sizeof(ByteCounter));
+    bufferPosition += sizeof(ByteCounter);
+    size += sizeof(ByteCounter);
+}
+
+
+ByteCounter TvgSaver::saveMemberData(const void* data, ByteCounter byteCnt)
+{
+    if (size + byteCnt > reserved) resizeBuffer(size + byteCnt);
+
+    memcpy(bufferPosition, data, byteCnt);
+    bufferPosition += byteCnt;
+    size += byteCnt;
+
+    return byteCnt;
+}
+
+
+ByteCounter TvgSaver::saveMember(IndicatorType ind, ByteCounter byteCnt, const void* data)
+{
+    ByteCounter blockByteCnt = sizeof(IndicatorType) + sizeof(ByteCounter) + byteCnt;
+
+    if (size + blockByteCnt > reserved) resizeBuffer(size + blockByteCnt);
+
+    memcpy(bufferPosition, &ind, sizeof(IndicatorType));
+    bufferPosition += sizeof(IndicatorType);
+    memcpy(bufferPosition, &byteCnt, sizeof(ByteCounter));
+    bufferPosition += sizeof(ByteCounter);
+    memcpy(bufferPosition, data, byteCnt);
+    bufferPosition += byteCnt;
+
+    size += blockByteCnt;
+
+    return blockByteCnt;
+}
+
+
+void TvgSaver::resizeBuffer(uint32_t newSize)
+{
+    // MGS - find more optimal alg ?
+    reserved += 100;
+    if (newSize > reserved) reserved += newSize - reserved + 100;
+
+    auto bufferOld = buffer;
+
+    buffer = static_cast<char*>(realloc(buffer, reserved));
```

Comment:
In tvg we don;t check the result of realloc, but here in some cases the buffer can be very big (many raw images) and the reallocation may fail - should I consider such situations?

I could save the buffer into the file after it reaches some size, but the problem is that I have to be able to move back within the buffer in order to complete the total number of bytes, after I write these bytes into the buffer. So if I clear the buffer I need to move within the file.  but it can be done of course.